File: backend/src/PisanteAtm.Application/Services/ProductService.cs
using Microsoft.Extensions.Logging;
using PisanteAtm.Application.Common;
using PisanteAtm.Application.DTOs.Product;
using PisanteAtm.Application.Interfaces;
using PisanteAtm.Domain.Entities;
using PisanteAtm.Domain.Interfaces;

namespace PisanteAtm.Application.Services;

public class ProductService : IProductService
{
    private readonly IUnitOfWork _uow;
    private readonly IFileStorageService _storage;
    private readonly ILogger<ProductService> _logger;

    public ProductService(IUnitOfWork uow, IFileStorageService storage, ILogger<ProductService> logger)
    {
        _uow = uow;
        _storage = storage;
        _logger = logger;
    }

    public async Task<PagedResult<ProductListDto>> GetPagedAsync(ProductQueryParams query, CancellationToken cancellationToken = default)
    {
        var (items, total) = await _uow.Products.GetPagedAsync(
            query.Page, query.PageSize,
            query.Search, query.CategoryId,
            query.MinPrice, query.MaxPrice,
            query.SortBy, query.Featured,
            cancellationToken);

        return new PagedResult<ProductListDto>
        {
            Items = items.Select(MapToListDto),
            TotalCount = total,
            Page = query.Page,
            PageSize = query.PageSize
        };
    }

    public async Task<ProductDto> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var product = await _uow.Products.GetWithDetailsAsync(id, cancellationToken)
            ?? throw AppException.NotFound("Product");
        return MapToDto(product);
    }

    public async Task<ProductDto> GetBySlugAsync(string slug, CancellationToken cancellationToken = default)
    {
        var product = await _uow.Products.GetBySlugAsync(slug, cancellationToken)
            ?? throw AppException.NotFound("Product");
        return MapToDto(product);
    }

    public async Task<IEnumerable<ProductListDto>> GetFeaturedAsync(int count = 8, CancellationToken cancellationToken = default)
    {
        var products = await _uow.Products.GetFeaturedAsync(count, cancellationToken);
        return products.Select(MapToListDto);
    }

    public async Task<ProductDto> CreateAsync(CreateProductDto dto, CancellationToken cancellationToken = default)
    {
        var category = await _uow.Categories.GetByIdAsync(dto.CategoryId, cancellationToken)
            ?? throw AppException.NotFound("Category");

        var slug = GenerateSlug(dto.Name);
        var exists = await _uow.Products.ExistsAsync(p => p.Slug == slug, cancellationToken);
        if (exists)
            slug = $"{slug}-{Guid.NewGuid().ToString()[..8]}";

        var product = new Product
        {
            Name = dto.Name,
            Slug = slug,
            Description = dto.Description,
            Price = dto.Price,
            DiscountPrice = dto.DiscountPrice,
            Brand = dto.Brand,
            CategoryId = dto.CategoryId,
            IsActive = dto.IsActive,
            IsFeatured = dto.IsFeatured
        };

        if (dto.Variants?.Any() == true)
        {
            var prefix = (dto.Name.Length > 4 ? dto.Name[..4] : dto.Name)
                .ToUpperInvariant().Replace(" ", "");
            product.Variants = dto.Variants.Select(v => new ProductVariant
            {
                Id = Guid.NewGuid(),
                Size = v.Size,
                Color = v.Color,
                SKU = !string.IsNullOrWhiteSpace(v.SKU)
                    ? v.SKU
                    : $"{prefix}-{v.Size}-{Guid.NewGuid().ToString()[..8]}".ToUpperInvariant(),
                StockQuantity = v.StockQuantity,
                PriceAdjustment = v.PriceAdjustment
            }).ToList();
        }

        await _uow.Products.AddAsync(product, cancellationToken);
        await _uow.SaveChangesAsync(cancellationToken);

        _logger.LogInformation("Product created: {Name} ({Id})", product.Name, product.Id);
        return await GetByIdAsync(product.Id, cancellationToken);
    }

    public async Task<ProductDto> UpdateAsync(Guid id, UpdateProductDto dto, CancellationToken cancellationToken = default)
    {
        var product = await _uow.Products.GetWithDetailsAsync(id, cancellationToken)
            ?? throw AppException.NotFound("Product");

        _ = await _uow.Categories.GetByIdAsync(dto.CategoryId, cancellationToken)
            ?? throw AppException.NotFound("Category");

        product.Name = dto.Name;
        product.Description = dto.Description;
        product.Price = dto.Price;
        product.DiscountPrice = dto.DiscountPrice;
        product.Brand = dto.Brand;
        product.CategoryId = dto.CategoryId;
        product.IsActive = dto.IsActive;
        product.IsFeatured = dto.IsFeatured;
        product.UpdatedAt = DateTime.UtcNow;

        if (dto.Variants?.Any() == true)
        {
            var existingVariantIds = product.Variants.Select(v => v.Id).ToList();
            foreach (var variantId in existingVariantIds)
            {
                var v = product.Variants.First(x => x.Id == variantId);
                product.Variants.Remove(v);
            }

            var prefix = (product.Name.Length > 4 ? product.Name[..4] : product.Name)
                .ToUpperInvariant().Replace(" ", "");
            foreach (var v in dto.Variants)
            {
                product.Variants.Add(new ProductVariant
                {
                    Id = Guid.NewGuid(),
                    ProductId = product.Id,
                    Size = v.Size,
                    Color = v.Color,
                    SKU = !string.IsNullOrWhiteSpace(v.SKU)
                        ? v.SKU
                        : $"{prefix}-{v.Size}-{Guid.NewGuid().ToString()[..8]}".ToUpperInvariant(),
                    StockQuantity = v.StockQuantity,
                    PriceAdjustment = v.PriceAdjustment
                });
            }
        }

        await _uow.SaveChangesAsync(cancellationToken);

        return await GetByIdAsync(product.Id, cancellationToken);
    }

    public async Task DeleteAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var product = await _uow.Products.GetByIdAsync(id, cancellationToken)
            ?? throw AppException.NotFound("Product");

        product.IsDeleted = true;
        product.UpdatedAt = DateTime.UtcNow;
        await _uow.Products.UpdateAsync(product, cancellationToken);
        await _uow.SaveChangesAsync(cancellationToken);
    }

    public async Task<string> UploadImageAsync(Guid productId, Stream imageStream, string fileName, bool isPrimary = false, CancellationToken cancellationToken = default)
    {
        var exists = await _uow.Products.ExistsAsync(p => p.Id == productId, cancellationToken);
        if (!exists)
            throw AppException.NotFound("Product");

        var url = await _storage.SaveAsync(imageStream, fileName, "products", cancellationToken);

        var existingImages = await _uow.ProductImages
            .FindAsync(i => i.ProductId == productId, cancellationToken);

        if (isPrimary)
        {
            foreach (var img in existingImages.Where(i => i.IsPrimary))
            {
                img.IsPrimary = false;
                await _uow.ProductImages.UpdateAsync(img, cancellationToken);
            }
        }

        await _uow.ProductImages.AddAsync(new ProductImage
        {
            ProductId = productId,
            Url = url,
            IsPrimary = isPrimary || !existingImages.Any(),
            DisplayOrder = existingImages.Count()
        }, cancellationToken);

        await _uow.SaveChangesAsync(cancellationToken);

        return _storage.GetPublicUrl(url);
    }

    public async Task DeleteImageAsync(Guid imageId, CancellationToken cancellationToken = default)
    {
        var images = await _uow.ProductImages.FindAsync(i => i.Id == imageId, cancellationToken);
        var image = images.FirstOrDefault()
            ?? throw AppException.NotFound("Image");

        await _storage.DeleteAsync(image.Url, cancellationToken);
        await _uow.ProductImages.DeleteAsync(image, cancellationToken);
        await _uow.SaveChangesAsync(cancellationToken);
    }

    public async Task<IEnumerable<ProductListDto>> GetLowStockAsync(int threshold = 5, CancellationToken cancellationToken = default)
    {
        var products = await _uow.Products.GetLowStockAsync(threshold, cancellationToken);
        return products.Select(MapToListDto);
    }

    private static string GenerateSlug(string name)
    {
        return name.ToLowerInvariant()
            .Replace(" ", "-")
            .Replace("ç", "c").Replace("ã", "a").Replace("â", "a")
            .Replace("á", "a").Replace("à", "a").Replace("é", "e")
            .Replace("ê", "e").Replace("í", "i").Replace("ó", "o")
            .Replace("ô", "o").Replace("õ", "o").Replace("ú", "u")
            .Replace("ü", "u")
            .Where(c => char.IsLetterOrDigit(c) || c == '-')
            .Aggregate("", (s, c) => s + c)
            .Trim('-');
    }

    private ProductDto MapToDto(Product p)
    {
        var imgUrl = (p.Images.FirstOrDefault(i => i.IsPrimary)
            ?? p.Images.OrderBy(i => i.DisplayOrder).FirstOrDefault())?.Url;
        return new(
            Id: p.Id,
            Name: p.Name,
            Slug: p.Slug,
            Description: p.Description,
            Price: p.Price,
            DiscountPrice: p.DiscountPrice,
            Brand: p.Brand,
            IsActive: p.IsActive,
            IsFeatured: p.IsFeatured,
            SalesCount: p.SalesCount,
            CategoryId: p.CategoryId,
            CategoryName: p.Category?.Name ?? string.Empty,
            Images: p.Images.OrderBy(i => i.DisplayOrder).Select(i => new ProductImageDto(
                i.Id, _storage.GetPublicUrl(i.Url), i.AltText, i.IsPrimary, i.DisplayOrder)).ToList(),
            Variants: p.Variants.Select(v => new ProductVariantDto(
                v.Id, v.Size, v.Color, v.SKU, v.StockQuantity, v.PriceAdjustment)).ToList(),
            AverageRating: p.Reviews.Any() ? p.Reviews.Average(r => r.Rating) : 0,
            ReviewCount: p.Reviews.Count,
            TotalStock: p.Variants.Sum(v => v.StockQuantity),
            CreatedAt: p.CreatedAt
        );
    }

    private ProductListDto MapToListDto(Product p)
    {
        var rawUrl = (p.Images.FirstOrDefault(i => i.IsPrimary)
            ?? p.Images.OrderBy(i => i.DisplayOrder).FirstOrDefault())?.Url;
        return new(
            Id: p.Id,
            Name: p.Name,
            Slug: p.Slug,
            Price: p.Price,
            DiscountPrice: p.DiscountPrice,
            Brand: p.Brand,
            IsFeatured: p.IsFeatured,
            PrimaryImageUrl: rawUrl != null ? _storage.GetPublicUrl(rawUrl) : null,
            CategoryName: p.Category?.Name ?? string.Empty,
            TotalStock: p.Variants.Sum(v => v.StockQuantity),
            AverageRating: p.Reviews.Any() ? p.Reviews.Average(r => r.Rating) : 0,
            SalesCount: p.SalesCount
        );
    }
}

File: frontend/src/services/productService.ts
import { supabase } from '@/lib/supabase'
import { slugify } from '@/lib/slugify'
import { toWebP } from '@/lib/imageUtils'
import type { PagedResult, Product, ProductList, ProductImage, ProductQueryParams } from '@/types'

interface RawVariant {
  id: string
  size: string
  color: string | null
  sku: string
  stock_quantity: number
}

interface RawImage {
  id: string
  url: string
  alt_text: string | null
  is_primary: boolean
  display_order: number
}

interface RawProduct {
  id: string
  category_id: string
  name: string
  slug: string
  description: string | null
  price: number
  discount_price: number | null
  brand: string | null
  is_active: boolean
  is_featured: boolean
  sales_count: number
  created_at: string
  categories: { name: string } | null
  product_images: RawImage[]
  product_variants: RawVariant[]
}

const SELECT = `
  *,
  categories(name),
  product_images(*),
  product_variants(*)
`

function toProductList(r: RawProduct): ProductList {
  const primary = r.product_images?.find((i) => i.is_primary) ?? r.product_images?.[0]
  return {
    id: r.id,
    name: r.name,
    slug: r.slug,
    price: r.price,
    discountPrice: r.discount_price ?? undefined,
    brand: r.brand ?? undefined,
    isFeatured: r.is_featured,
    primaryImageUrl: primary?.url,
    categoryName: r.categories?.name ?? '',
    totalStock: r.product_variants?.reduce((s, v) => s + v.stock_quantity, 0) ?? 0,
    averageRating: 0,
    salesCount: r.sales_count,
  }
}

function toProduct(r: RawProduct): Product {
  return {
    id: r.id,
    name: r.name,
    slug: r.slug,
    description: r.description ?? undefined,
    price: r.price,
    discountPrice: r.discount_price ?? undefined,
    brand: r.brand ?? undefined,
    isActive: r.is_active,
    isFeatured: r.is_featured,
    salesCount: r.sales_count,
    categoryId: r.category_id,
    categoryName: r.categories?.name ?? '',
    images: (r.product_images ?? [])
      .sort((a, b) => a.display_order - b.display_order)
      .map((img) => ({
        id: img.id,
        url: img.url,
        altText: img.alt_text ?? undefined,
        isPrimary: img.is_primary,
        displayOrder: img.display_order,
      })),
    variants: (r.product_variants ?? []).map((v) => ({
      id: v.id,
      size: v.size,
      color: v.color ?? undefined,
      sku: v.sku,
      stockQuantity: v.stock_quantity,
      priceAdjustment: 0,
    })),
    averageRating: 0,
    reviewCount: 0,
    totalStock: r.product_variants?.reduce((s, v) => s + v.stock_quantity, 0) ?? 0,
    createdAt: r.created_at,
  }
}

export const productService = {
  getAll: async (params?: ProductQueryParams): Promise<PagedResult<ProductList>> => {
    const page = params?.page ?? 1
    const pageSize = params?.pageSize ?? 12
    const from = (page - 1) * pageSize
    const to = from + pageSize - 1

    let q = supabase.from('products').select(SELECT, { count: 'exact' })

    if (params?.search) q = q.ilike('name', `%${params.search}%`)
    if (params?.categoryId) q = q.eq('category_id', params.categoryId)
    if (params?.minPrice !== undefined) q = q.gte('price', params.minPrice)
    if (params?.maxPrice !== undefined) q = q.lte('price', params.maxPrice)
    if (params?.featured) q = q.eq('is_featured', true)

    switch (params?.sortBy) {
      case 'price_asc': q = q.order('price', { ascending: true }); break
      case 'price_desc': q = q.order('price', { ascending: false }); break
      case 'sales': q = q.order('sales_count', { ascending: false }); break
      default: q = q.order('created_at', { ascending: false })
    }

    const { data, error, count } = await q.range(from, to)
    if (error) throw error

    const totalCount = count ?? 0
    const totalPages = Math.ceil(totalCount / pageSize)
    return {
      items: (data as unknown as RawProduct[]).map(toProductList),
      totalCount,
      page,
      pageSize,
      totalPages,
      hasNextPage: page < totalPages,
      hasPreviousPage: page > 1,
    }
  },

  getFeatured: async (count = 8): Promise<ProductList[]> => {
    const { data, error } = await supabase
      .from('products')
      .select(SELECT)
      .eq('is_featured', true)
      .order('sales_count', { ascending: false })
      .limit(count)
    if (error) throw error
    return (data as unknown as RawProduct[]).map(toProductList)
  },

  getById: async (id: string): Promise<Product> => {
    const { data, error } = await supabase
      .from('products')
      .select(SELECT)
      .eq('id', id)
      .single()
    if (error) throw error
    return toProduct(data as unknown as RawProduct)
  },

  getBySlug: async (slug: string): Promise<Product> => {
    const { data, error } = await supabase
      .from('products')
      .select(SELECT)
      .eq('slug', slug)
      .single()
    if (error) throw error
    return toProduct(data as unknown as RawProduct)
  },

  getLowStock: async (threshold = 5): Promise<ProductList[]> => {
    const { data, error } = await supabase
      .from('products')
      .select(SELECT)
      .order('name')
    if (error) throw error
    return (data as unknown as RawProduct[])
      .filter((p) => (p.product_variants ?? []).some((v) => v.stock_quantity <= threshold))
      .map(toProductList)
  },

  create: async (payload: {
    name: string
    description?: string
    price: number
    discountPrice?: number
    brand?: string
    categoryId: string
    isActive: boolean
    isFeatured: boolean
    variants: Array<{ size: string; color?: string; sku: string; stockQuantity: number }>
  }): Promise<Product> => {
    const { data: product, error } = await supabase
      .from('products')
      .insert({
        category_id: payload.categoryId,
        name: payload.name,
        slug: slugify(payload.name),
        description: payload.description ?? null,
        price: payload.price,
        discount_price: payload.discountPrice ?? null,
        brand: payload.brand ?? null,
        is_active: payload.isActive,
        is_featured: payload.isFeatured,
      })
      .select('id')
      .single()
    if (error) throw error

    if (payload.variants.length > 0) {
      const { error: varErr } = await supabase.from('product_variants').insert(
        payload.variants.map((v) => ({
          product_id: product.id,
          size: v.size,
          color: v.color ?? null,
          sku: v.sku,
          stock_quantity: v.stockQuantity,
        })),
      )
      if (varErr) throw varErr
    }

    return productService.getById(product.id)
  },

  update: async (
    id: string,
    payload: {
      name?: string
      description?: string
      price?: number
      discountPrice?: number
      brand?: string
      categoryId?: string
      isActive?: boolean
      isFeatured?: boolean
      variants?: Array<{ id?: string; size: string; color?: string; sku: string; stockQuantity: number }>
    },
  ): Promise<Product> => {
    const updates: Record<string, unknown> = { updated_at: new Date().toISOString() }
    if (payload.name !== undefined) { updates.name = payload.name; updates.slug = slugify(payload.name) }
    if (payload.description !== undefined) updates.description = payload.description || null
    if (payload.price !== undefined) updates.price = payload.price
    if ('discountPrice' in payload) updates.discount_price = payload.discountPrice ?? null
    if (payload.brand !== undefined) updates.brand = payload.brand || null
    if (payload.categoryId !== undefined) updates.category_id = payload.categoryId
    if (payload.isActive !== undefined) updates.is_active = payload.isActive
    if (payload.isFeatured !== undefined) updates.is_featured = payload.isFeatured

    const { error } = await supabase.from('products').update(updates).eq('id', id)
    if (error) throw error

    if (payload.variants !== undefined) {
      await supabase.from('product_variants').delete().eq('product_id', id)
      if (payload.variants.length > 0) {
        const { error: varErr } = await supabase.from('product_variants').insert(
          payload.variants.map((v) => ({
            product_id: id,
            size: v.size,
            color: v.color ?? null,
            sku: v.sku,
            stock_quantity: v.stockQuantity,
          })),
        )
        if (varErr) throw varErr
      }
    }

    return productService.getById(id)
  },

  delete: async (id: string): Promise<void> => {
    // Busca URLs das imagens para limpar o storage após deletar o registro
    const { data: images } = await supabase
      .from('product_images')
      .select('url')
      .eq('product_id', id)

    const { error } = await supabase.from('products').delete().eq('id', id)
    if (error) throw error

    if (images?.length) {
      const paths = images
        .map((img) => img.url.split('/product-images/')[1])
        .filter(Boolean)
      if (paths.length) {
        await supabase.storage.from('product-images').remove(paths)
      }
    }
  },

  uploadImage: async (productId: string, file: File, isPrimary = false): Promise<ProductImage> => {
    const blob = await toWebP(file)
    const path = `${productId}/${crypto.randomUUID()}.webp`

    const { error: uploadErr } = await supabase.storage
      .from('product-images')
      .upload(path, blob, { contentType: 'image/webp' })
    if (uploadErr) throw uploadErr

    const { data: { publicUrl } } = supabase.storage
      .from('product-images')
      .getPublicUrl(path)

    if (isPrimary) {
      await supabase
        .from('product_images')
        .update({ is_primary: false })
        .eq('product_id', productId)
    }

    const { data, error: dbErr } = await supabase
      .from('product_images')
      .insert({ product_id: productId, url: publicUrl, is_primary: isPrimary, display_order: 0 })
      .select()
      .single()
    if (dbErr) throw dbErr

    return {
      id: data.id,
      url: data.url,
      altText: data.alt_text ?? undefined,
      isPrimary: data.is_primary,
      displayOrder: data.display_order,
    }
  },

  deleteImage: async (imageId: string): Promise<void> => {
    const { data: img } = await supabase
      .from('product_images')
      .select('url')
      .eq('id', imageId)
      .single()

    const { error } = await supabase.from('product_images').delete().eq('id', imageId)
    if (error) throw error

    if (img?.url) {
      const storagePath = img.url.split('/product-images/')[1]
      if (storagePath) {
        await supabase.storage.from('product-images').remove([storagePath])
      }
    }
  },
}
